apps/omg_child_chain/lib/omg_child_chain/api/blocks_cache/storage.ex

```diff
@@ -0,0 +1,62 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.ChildChain.API.BlocksCache.Storage do
+  @moduledoc """
+  Logic of the service to serve freshest blocks quickly.
+  """
+  alias OMG.Block
+  alias OMG.DB
+
+  @doc """
+  The idea behind having another lookup in ETS is that once a 
+  fresh block is published every watcher is racing to get that block.
+  The first read in BlocksCache would be a miss, and the request message would go sit in the
+  gen server queue. But the first Watcher that requested the block would insert the block in
+  ETS. We're trying to protect RocksDB single server.
```

Comment:
I'm not really following here. There's no sleep between the two tries or anything that would give enough time for the block to be inserted 🤔 It's basically going to try retrieving it again after <1 ms, which won't be enough. Unless I'm missing something, I don't think this brings any benefits with the current implementation.

Reply:
assuming `ets.insert` is a synchronous operation (I actually don't know is this true or not 😛)
I think it is guarded by the mailbox of `API.BlocksCache` so it is talking on a higher level and not really in this module (I guess).

hmm...but now it introduce another singleton process assumption, which have seen to be broken when we want to scale to multi ch-ch nodes with our current style of deployment. 🤔

Reply:
The reason why this second lookup speeds things us is because it elevates pressure from requesting data from RocksDB. If you target a small number of blocks (imagine a new block published by childchain). All watchers try to get that specific block. `BlocksCache ` is a process maintaining a queue of those requests. And the mailbox is processesed synchronusly. The first time you get a request for block `asdf` you would insert that block in the ETS cache and respond, every request after the first one can get the block from the ETS instead of RocksDB. So there's that.